### PR TITLE
replace `toaster` with new `useToaster` api

### PR DESCRIPTION
**Pull request by mayank99**
## Changes

The `Toaster` class instance been replaced with a `useToaster` hook which returns an object that has properties matching the public functions in the old api.

Only api change needed from user side is:
```diff
- import { toaster } from '@itwin/itwinui-react';
+ import { useToaster } from '@itwin/itwinui-react';

  export default () => {
+   const toaster = useToaster();

    return <button onClick={() => toaster.positive('hello')}>show</button>
  };
```

### How it works

- The `useToaster` hook simply dispatches reducer actions.
- The reducer state is stored in `<ToastProvider>`, which wraps the entire tree inside `<ThemeProvider>`.
- `<ThemeProvider>` renders a `<Toaster>` which reads the state and renders `<Toast>`s accordingly.

~~TODO: Will work on unit tests~~ Fixed.

## Testing

Tested in vite playground and storybook.

Everything seems to work well. 🎉

## Docs

Added changeset.

~~Need to fix website example.~~ Fixed via #1353.

### CODE REVIEW

File: packages/itwinui-react/src/core/Toast/Toaster.tsx [hunk 1]
```diff
@@ -38,140 +35,141 @@ export type ToastOptions = Omit<
...
   };
+};
 
-  /**
-   * Set global Toaster settings for toasts order and placement.
-   * Settings will be applied to new toasts on the page.
-   */
-  public setSettings(newSettings: ToasterSettings) {
-    newSettings.placement ??= this.settings.placement;
-    newSettings.order ??= newSettings.placement?.startsWith('bottom')
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Toast/Toaster.tsx:96]: Was this functionality removed on purpose?

File: packages/itwinui-react/src/core/Toast/Toaster.tsx [hunk 2]
```diff
@@ -38,140 +35,141 @@ export type ToastOptions = Omit<
...
-  }
+// ----------------------------------------------------------------------------
 
-  public warning(content: React.ReactNode, options?: ToastOptions) {
-    return this.createToast(content, 'warning', options);
-  }
+export const ToastProvider = ({ children }: { children: React.ReactNode }) => {
+  const [toasterState, dispatch] = React.useReducer(toastReducer, {
+    toasts: [],
+    placement: 'top',
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Toast/Toaster.tsx:91]: Why placement is repeated two times?

File: packages/itwinui-react/src/core/Toast/Toaster.tsx [hunk 3]
```diff
@@ -18,13 +15,13 @@ export type ToasterSettings = {
 
    * When `placement` is set to a top value, order defaults to 'descending', otherwise 'ascending'.
    */
-  order?: 'descending' | 'ascending';
+  order: 'descending' | 'ascending';
```
Comment (gretanausedaite) [inline packages/itwinui-react/src/core/Toast/Toaster.tsx:18]: Both setting props are mandatory now?

**Review by gretanausedaite — COMMENTED** [review]
Couple questions. Other than that component looks much simpler!

File: packages/itwinui-react/src/core/Toast/Toaster.tsx (hunk 2 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Toast/Toaster.tsx]: removed

File: packages/itwinui-react/src/core/Toast/Toaster.tsx (hunk 3 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Toast/Toaster.tsx:18]: no, it's using `Partial` in the dispatch action (line 165)

File: packages/itwinui-react/src/core/Toast/Toaster.tsx (hunk 1 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Toast/Toaster.tsx:96]: hmm, not intentionally. But I'm thinking about it - previously we allowed `order` to be optional, whereas now it gets a default value. We could introduce a third option `auto` and make that the default?

**Review by gretanausedaite — APPROVED** [review]
LGTM.

**Pull request merged**